### PR TITLE
[doc] Improve dev_install for Windows and LLVM 10

### PR DESCRIPTION
**Pull request by yuanming-hu**
Related issue = #655

- [[Click here for the format server]](http://kun.csail.mit.edu:31415/)
- [[Click here for coverage report]](http://codecov.io/gh/taichi-dev/taichi/)

### CODE REVIEW

File: docs/dev_install.rst [hunk 1]
```diff
@@ -11,48 +11,83 @@ For precise build instructions on Windows, please check out `appveyor.yml <https
...
+        cd cfe-8.0.1.src
+        mkdir build
+        cd build
+        cmake ..
+        make -j 8
+        sudo make install
+
+
+- Make sure you have LLVM 8.0.1/10.0.0. Note that Taichi uses a customized LLVM so the pre-built binaries from the LLVM official website or other sources probably doesn't work.
+  Here we provide LLVM 8.0.1 binaries customized for Taichi, which may or may not work depending on your system environment:
```
Comment (TH3CHARLie) [inline docs/dev_install.rst:50]: Although not really related to this PR, what kind of customizations happen in the pre-built binaries? I believe this is related to tasks in #655.

File: docs/dev_install.rst [hunk 2]
```diff
@@ -11,48 +11,83 @@ For precise build instructions on Windows, please check out `appveyor.yml <https
...
 
   .. code-block:: bash
 
-    python3 -m pip install --user setuptools astor pybind11 numpy
-    python3 -m pip install --user pytest pytest-rerunfailures pytest-xdist pytest-cov
-    python3 -m pip install --user GitPython coverage autograd Pillow yapf
-    python3 -m pip install --user scipy psutil opencv-python astpretty
+    python3 -m pip install --user setuptools astpretty astor pybind11
+    python3 -m pip install --user pytest pytest-rerunfailures pytest-xdist yapf Pillow
+    python3 -m pip install --user numpy GitPython coverage colorama autograd
```
Comment (archibate) [inline docs/dev_install.rst:24]:
```suggestion
    python3 -m pip install --user setuptools astpretty astor pybind11 Pillow
    python3 -m pip install --user pytest pytest-rerunfailures pytest-xdist yapf
    python3 -m pip install --user numpy GitPython coverage colorama autograd
```

File: docs/dev_install.rst [hunk 3]
```diff
@@ -11,48 +11,83 @@ For precise build instructions on Windows, please check out `appveyor.yml <https
...
+
+- Make sure you have LLVM 8.0.1/10.0.0. Note that Taichi uses a customized LLVM so the pre-built binaries from the LLVM official website or other sources probably doesn't work.
+  Here we provide LLVM 8.0.1 binaries customized for Taichi, which may or may not work depending on your system environment:
+  `Linux <https://github.com/yuanming-hu/taichi_assets/releases/download/llvm8/taichi-llvm-8.0.1-linux-x64.zip>`_,
+  `OS X <https://github.com/yuanming-hu/taichi_assets/releases/download/llvm8/taichi-llvm-8.0.1.zip>`_,
+  `Windows <https://github.com/yuanming-hu/taichi_assets/releases/download/llvm8/taichi-llvm-8.0.1-msvc2017.zip>`_.
+
+   If the downloaded LLVM does not work, please build from source:
+
+  - On Linux or OS X:
```
Comment (archibate) [inline docs/dev_install.rst:57]:
```suggestion
  * On Linux or OS X:
```

Comment (archibate) [inline docs/dev_install.rst:57]:
```suggestion
  * On Linux or OS X:
```

File: docs/dev_install.rst [hunk 4]
```diff
@@ -11,48 +11,83 @@ For precise build instructions on Windows, please check out `appveyor.yml <https
...
+        tar xvJf llvm-8.0.1.src.tar.xz
+        cd llvm-8.0.1.src
+        mkdir build
+        cd build
+        cmake .. -DLLVM_ENABLE_RTTI:BOOL=ON -DBUILD_SHARED_LIBS:BOOL=OFF -DCMAKE_BUILD_TYPE=Release -DLLVM_TARGETS_TO_BUILD="X86;NVPTX" -DLLVM_ENABLE_ASSERTIONS=ON
+        # If you are building on NVIDIA Jetson TX2, use -DLLVM_TARGETS_TO_BUILD="ARM;NVPTX"
+        make -j 8
+        sudo make install
+
+  - On Windows:
```
Comment (archibate) [inline docs/dev_install.rst:71]:
```suggestion
  * On Windows:
```

File: docs/dev_install.rst [hunk 5]
```diff
@@ -11,48 +11,83 @@ For precise build instructions on Windows, please check out `appveyor.yml <https
...
+
+    Then open ``LLVM.sln`` and use Visual Studio 2017+ to build. Please make sure you are using the ``Release`` configuration. After building the ``INSTALL`` project (under folder ``CMakePredefinedTargets`` in the Solution Explorer window).
+    After the build is complete, find your LLVM binaries and headers in ``build/installed``.
+    If you use MSVC 2019, **make sure you use C++17** for the ``INSTALL`` project.
+
+    Please add ``build/installed/bin`` to ``PATH``.
+    Later, when you build Taichi using ``CMake``, set ``LLVM_DIR`` to ``build/installed/lib/cmake/llvm``.
+
+- On Windows, if you use the pre-built LLVM for Taichi, please add ``$LLVM_FOLDER/bin`` to ``PATH``.
+  Later, when you build Taichi using ``CMake``, set ``LLVM_DIR`` to ``$LLVM_FOLDER/lib/cmake/llvm``.
```
Comment (archibate) [inline docs/dev_install.rst:89]:
```suggestion
- On Windows, if you use the pre-built LLVM for Taichi, please:
   1. Add ``<llvm folder>/bin`` to ``%PATH%``.
   2. Set ``%LLVM_DIR%`` to ``<llvm folder>/lib/cmake/llvm``.
```

File: docs/dev_install.rst [hunk 6]
```diff
@@ -11,48 +11,83 @@ For precise build instructions on Windows, please check out `appveyor.yml <https
...
+
+      # LLVM 10.0.0 + MSVC 2019
+      cmake .. -G"Visual Studio 16 2019" -A x64  -DLLVM_ENABLE_RTTI:BOOL=ON -DBUILD_SHARED_LIBS:BOOL=OFF -DCMAKE_BUILD_TYPE=Release -DLLVM_TARGETS_TO_BUILD="X86;NVPTX" -DLLVM_ENABLE_ASSERTIONS=ON -Thost=x64 -DLLVM_BUILD_TESTS:BOOL=OFF -DCMAKE_INSTALL_PREFIX=installed
+
+    Then open ``LLVM.sln`` and use Visual Studio 2017+ to build. Please make sure you are using the ``Release`` configuration. After building the ``INSTALL`` project (under folder ``CMakePredefinedTargets`` in the Solution Explorer window).
+    After the build is complete, find your LLVM binaries and headers in ``build/installed``.
+    If you use MSVC 2019, **make sure you use C++17** for the ``INSTALL`` project.
+
+    Please add ``build/installed/bin`` to ``PATH``.
+    Later, when you build Taichi using ``CMake``, set ``LLVM_DIR`` to ``build/installed/lib/cmake/llvm``.
```
Comment (archibate) [inline docs/dev_install.rst:86]:
```suggestion
    Then configure the environment variables:
     1. Add ``build/installed/bin`` to ``%PATH%``.
     2. Set ``%LLVM_DIR%`` to ``build/installed/lib/cmake/llvm``.
```

File: docs/dev_install.rst (hunk 1 above)
Comment (yuanming-hu) [reply docs/dev_install.rst:50]: Basically the additional CMake options: https://github.com/taichi-dev/taichi/pull/1202/files#diff-97322f732b832df07874b286a638a2a1R66

I guess RTTI is what leads to incompatibilities (although I haven't confirmed).

File: docs/dev_install.rst [hunk 7]
```diff
@@ -11,48 +11,83 @@ For precise build instructions on Windows, please check out `appveyor.yml <https
...
 
-* (If on Ubuntu) Execute ``sudo apt install libtinfo-dev clang-8``. (``clang-7`` should work as well).
+- Make sure you have ``clang`` with version >= 7
 
-* (If on other Linux distributions) Please build clang 8.0.1 from source:
+  * On Windows: Download ``clang-8`` via `this link <https://releases.llvm.org/8.0.0/LLVM-8.0.0-win64.exe>`_.
+    Make sure you add the ``bin`` folder containing ``clang.exe`` to the ``PATH`` environment variable.
 
-  .. code-block:: bash
+  * On OS X: you don't need to do anything.
```
[the remark this answers is not in the document]
Comment (yuanming-hu) [reply docs/dev_install.rst:31]: I see, thanks for mentioning that! Could you push to this PR about this or open a future PR on this topic?

File: docs/dev_install.rst [hunk 8]
```diff
@@ -11,48 +11,83 @@ For precise build instructions on Windows, please check out `appveyor.yml <https
...
+
+      # LLVM 8.0.1 + MSVC 2017
+      cmake .. -G"Visual Studio 15 2017 Win64"  -DLLVM_ENABLE_RTTI:BOOL=ON -DBUILD_SHARED_LIBS:BOOL=OFF -DCMAKE_BUILD_TYPE=Release -DLLVM_TARGETS_TO_BUILD="X86;NVPTX" -DLLVM_ENABLE_ASSERTIONS=ON -Thost=x64 -DLLVM_BUILD_TESTS:BOOL=OFF -DCMAKE_INSTALL_PREFIX=installed
+
+      # LLVM 10.0.0 + MSVC 2019
+      cmake .. -G"Visual Studio 16 2019" -A x64  -DLLVM_ENABLE_RTTI:BOOL=ON -DBUILD_SHARED_LIBS:BOOL=OFF -DCMAKE_BUILD_TYPE=Release -DLLVM_TARGETS_TO_BUILD="X86;NVPTX" -DLLVM_ENABLE_ASSERTIONS=ON -Thost=x64 -DLLVM_BUILD_TESTS:BOOL=OFF -DCMAKE_INSTALL_PREFIX=installed
+
+    Then open ``LLVM.sln`` and use Visual Studio 2017+ to build. Please make sure you are using the ``Release`` configuration. After building the ``INSTALL`` project (under folder ``CMakePredefinedTargets`` in the Solution Explorer window).
+    After the build is complete, find your LLVM binaries and headers in ``build/installed``.
+    If you use MSVC 2019, **make sure you use C++17** for the ``INSTALL`` project.
```
Comment (archibate) [inline docs/dev_install.rst:83]:
```suggestion
    1. Then open and build ``LLVM.sln`` with Visual Studio 2017+. Make sure you are using the ``Release`` configuration.
    2. Then build the ``INSTALL`` project under folder ``CMakePredefinedTargets`` in the Solution Explorer window.
    
    .. note::
    
        After the build is complete, you may find your LLVM binaries and headers in ``build/installed``.
        If you use MSVC 2019, **make sure you use C++17** for the ``INSTALL`` project.
```

**Comment by Eydcao** [comment]
Hi Yuanming,

Should we also address the multiple instances of MSBUILD.exe on Windows (first commit in  #1021)? Like a simple note in the ymal file? This stuck me for a long time once.

**Comment by yuanming-hu** [comment]
> Hi Yuanming,
> 
> Should we also address the multiple instances of MSBUILD.exe on Windows (first commit in #1021)? Like a simple note in the ymal file? This stuck me for a long time once.

Sure, please do that in #1021. Sorry about getting you stuck!

File: docs/dev_install.rst (hunk 7 above)
[the remark this answers is not in the document]
Comment (TH3CHARLie) [reply docs/dev_install.rst:31]: Let's keep this PR related to Windows. 

Meanwhile, when testing LLVM 10 under macOS, I'll also test its compatibility with the currently working(Xcode 11.3) and the latest toolchains. Then we open a PR to add all the notes regarding installation on macOS, just like this one. But hopefully, with the updates, we can achieve the ideal "you don't need to do anything".

**Comment by Eydcao** [comment]
> > Hi Yuanming,
> > Should we also address the multiple instances of MSBUILD.exe on Windows (first commit in #1021)? Like a simple note in the ymal file? This stuck me for a long time once.
> 
> Sure, please do that in #1021. Sorry about getting you stuck!

Sure thing I will handle the conflict there.~ Thanks LGTM!

File: docs/dev_install.rst (hunk 5 above)
Comment (yuanming-hu) [reply docs/dev_install.rst:89]: Actually, the `LLVM_DIR` is a CMake option instead of an environment variable.

File: docs/dev_install.rst (hunk 8 above)
Comment (yuanming-hu) [reply docs/dev_install.rst:83]: Thanks! I'll use a bulleted list, but the `note` does not quite necessary here.

**Pull request merged**